src/message.rs

```diff
@@ -28,15 +31,129 @@ fn _message_prettify(message: CString, comment_char: Option<u8>) -> Result<Strin
     Ok(ret.as_str().unwrap().to_string())
 }
 
+/// Collection of trailer key–value pairs.
+///
+/// Use `iter()` to get access to the values.
+pub struct MessageTrailers<'pair> {
+    raw: raw::git_message_trailer_array,
+    _marker: marker::PhantomData<&'pair c_char>,
+}
+
+impl<'pair> MessageTrailers<'pair> {
+    fn new() -> MessageTrailers<'pair> {
+        crate::init();
+        unsafe {
+            Binding::from_raw(&mut raw::git_message_trailer_array {
+                trailers: ptr::null_mut(),
+                count: 0,
+                _trailer_block: ptr::null_mut(),
+            } as *mut _)
+        }
+    }
+    /// Create a borrowed iterator.
+    pub fn iter(&'pair self) -> MessageTrailersIterator<'pair> {
+        MessageTrailersIterator {
+            trailers: self,
+            counter: 0,
+        }
+    }
+    /// The number of trailer key–value pairs.
+    pub fn len(&self) -> usize {
+        self.raw.count
+    }
+}
+
+impl<'pair> Drop for MessageTrailers<'pair> {
+    fn drop(&mut self) {
+        unsafe {
+            raw::git_message_trailer_array_free(&mut self.raw);
+        }
+    }
+}
+
+impl<'pair> Binding for MessageTrailers<'pair> {
+    type Raw = *mut raw::git_message_trailer_array;
+    unsafe fn from_raw(raw: *mut raw::git_message_trailer_array) -> MessageTrailers<'pair> {
+        MessageTrailers {
+            raw: *raw,
+            _marker: marker::PhantomData,
+        }
+    }
+    fn raw(&self) -> *mut raw::git_message_trailer_array {
+        &self.raw as *const _ as *mut _
+    }
+}
+
+struct Trailer<'pair> {
+    key: *const c_char,
+    value: *const c_char,
+    _marker: marker::PhantomData<&'pair c_char>,
+}
+
+impl<'pair> Trailer<'pair> {
+    fn to_str_tuple(self) -> (&'pair str, &'pair str) {
+        unsafe {
+            let key = CStr::from_ptr(self.key).to_str().unwrap();
+            let value = CStr::from_ptr(self.value).to_str().unwrap();
+            (key, value)
+        }
+    }
+}
+
+/// A borrowed iterator.
+pub struct MessageTrailersIterator<'pair> {
+    trailers: &'pair MessageTrailers<'pair>,
+    counter: usize,
+}
+
+impl<'pair> Iterator for MessageTrailersIterator<'pair> {
+    type Item = (&'pair str, &'pair str);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.counter == self.trailers.raw.count {
+            None
+        } else {
+            unsafe {
+                let addr = self.trailers.raw.trailers.wrapping_add(self.counter);
+                self.counter += 1;
+                Some(
+                    Trailer {
+                        key: (*addr).key,
+                        value: (*addr).value,
+                        _marker: marker::PhantomData,
+                    }
+                    .to_str_tuple(),
+                )
+            }
+        }
+    }
+}
```

Comment:
Could you add implementations for `size_hint` and `ExactSizeIterator` here as well? (which would also necessitate `DoubleEndedIterator`, and the implementations should be relatively easy if the `counter` field is changed to a `Range`)